paimon-core/src/main/java/org/apache/paimon/utils/BranchManager.java

```diff
@@ -0,0 +1,160 @@
+import org.apache.paimon.fs.FileIO;
+import org.apache.paimon.fs.Path;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+import static org.apache.paimon.schema.SchemaManager.SCHEMA_PREFIX;
+import static org.apache.paimon.utils.Preconditions.checkArgument;
+import static org.apache.paimon.utils.SnapshotManager.SNAPSHOT_PREFIX;
+import static org.apache.paimon.utils.TagManager.TAG_PREFIX;
+
+/** Manager for {@code Branch}. */
+public class BranchManager {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BranchManager.class);
+
+    private static final String BRANCH_PREFIX = "branch-";
+
+    private final FileIO fileIO;
+    private final Path tablePath;
+
+    public BranchManager(FileIO fileIO, Path path) {
+        this.fileIO = fileIO;
+        this.tablePath = path;
+    }
+
+    /** Return the root Directory of branch. */
+    public Path branchDirectory() {
+        return new Path(tablePath + "/branch");
+    }
+
+    /** Return the path of a branch. */
+    public String getBranchPath(String branchName) {
+        return tablePath + "/branch/" + BRANCH_PREFIX + branchName;
+    }
+
+    public Path branchPath(String branchName) {
+        return new Path(getBranchPath(branchName));
+    }
+
+    public Path tagPath(String tagName) {
+        return new Path(tablePath + "/tag/" + TAG_PREFIX + tagName);
+    }
+
+    public Path branchTagPath(String branchName, String tagName) {
+        return new Path(getBranchPath(branchName) + "/tag/" + TAG_PREFIX + tagName);
+    }
+
+    public Path snapshotPath(long snapshotId) {
+        return new Path(tablePath + "/snapshot/" + SNAPSHOT_PREFIX + snapshotId);
+    }
+
+    public Path branchSnapshotPath(String branchName, long snapshotId) {
+        return new Path(getBranchPath(branchName) + "/snapshot/" + SNAPSHOT_PREFIX + snapshotId);
+    }
+
+    public Path schemaPath(long schemaId) {
+        return new Path(tablePath + "/schema/" + SCHEMA_PREFIX + schemaId);
+    }
+
+    public Path branchSchemaPath(String branchName, long schemaId) {
+        return new Path(getBranchPath(branchName) + "/schema/" + SCHEMA_PREFIX + schemaId);
+    }
+
+    public void createBranch(String branchName, String tagName) {
+        checkArgument(!StringUtils.isBlank(branchName), "Branch name '%s' is blank.", branchName);
+        checkArgument(!branchExists(branchName), "Branch name '%s' already exists.", branchName);
+        checkArgument(tagExists(tagName), "Tag name '%s' not exists.", tagName);
+        checkArgument(
+                !branchName.chars().allMatch(Character::isDigit),
+                "Branch name cannot be pure numeric string but is '%s'.",
+                branchName);
+
+        Snapshot snapshot = taggedSnapshot(tagName);
+
+        try {
+            // Copy the corresponding tag, snapshot and schema files into the branch directory
+            fileIO.copyFileUtf8(tagPath(tagName), branchTagPath(branchName, tagName));
```

Comment:
Why should we copy tag file to branch directory?

Reply:
From design:
![image](https://github.com/apache/incubator-paimon/assets/45089228/b421e6be-94c4-4d08-8392-6ce31a0eec09)

In addition, I think it makes sense to copy tag. The new branch will make it easier for users to use tag. What do you think about it？